In [303]:
import pandas as pd
pd.options.display.max_columns = 200
import numpy as np


def prettyTable(dataframe, column):
    z = pd.DataFrame(columns = [dataframe[column].value_counts(True).index, round(dataframe[column].value_counts(True) * 100, 2).astype(str) + '%', dataframe[column].value_counts()]).T
    return z

In [276]:
# loans_full_schema = pd.read_csv('loans_full_schema.csv')
# print(loans_full_schema['loan_status'].value_counts(dropna = False))
# loans_full_schema

In [277]:
# accepted_2007_to_2018Q4 = pd.read_csv('accepted_2007_to_2018Q4.csv', nrows = 10000)
# accepted_2007_to_2018Q4 = accepted_2007_to_2018Q4[sorted(accepted_2007_to_2018Q4.columns.tolist())]
# print(accepted_2007_to_2018Q4['loan_status'].value_counts(dropna = False))
# accepted_2007_to_2018Q4

## Cleaning and Initial EDA

In [278]:
cr_loan2 = pd.read_csv('cr_loan2.csv')
cr_loan2

person_age  person_income person_home_ownership  person_emp_length  \
0              22          59000                  RENT              123.0   
1              21           9600                   OWN                5.0   
2              25           9600              MORTGAGE                1.0   
3              23          65500                  RENT                4.0   
4              24          54400                  RENT                8.0   
...           ...            ...                   ...                ...   
32576          57          53000              MORTGAGE                1.0   
32577          54         120000              MORTGAGE                4.0   
32578          65          76000                  RENT                3.0   
32579          56         150000              MORTGAGE                5.0   
32580          66          42000                  RENT                2.0   

           loan_intent loan_grade  loan_amnt  loan_int_rate  loan_status  \
0             PERSONAL          D      35000          16.02            1   
1            EDUCATION          B       1000          11.14            0   
2              MEDICAL          C       5500          12.87            1   
3              MEDICAL          C      35000          15.23            1   
4              MEDICAL          C      35000          14.27            1   
...                ...        ...        ...            ...          ...   
32576         PERSONAL          C       5800          13.16            0   
32577         PERSONAL          A      17625           7.49            0   
32578  HOMEIMPROVEMENT          B      35000          10.99            1   
32579         PERSONAL          B      15000          11.48            0   
32580          MEDICAL          B       6475           9.99            0   

       loan_percent_income cb_person_default_on_file  \
0                     0.59                         Y   
1                     0.10                         N   
2                     0.57                         N   
3                     0.53                         N   
4                     0.55                         Y   
...                    ...                       ...   
32576                 0.11                         N   
32577                 0.15                         N   
32578                 0.46                         N   
32579                 0.10                         N   
32580                 0.15                         N   

       cb_person_cred_hist_length  
0                               3  
1                               2  
2                               3  
3                               2  
4                               4  
...                           ...  
32576                          30  
32577                          19  
32578                          28  
32579                          26  
32580                          30  

[32581 rows x 12 columns]

In [279]:
print(cr_loan2['loan_status'].value_counts(True, dropna = False))
cr_loan2['loan_status'].value_counts(dropna = False)

0    0.781836
1    0.218164
Name: loan_status, dtype: float64


0    25473
1     7108
Name: loan_status, dtype: int64

In [280]:
df = cr_loan2.copy()
df.isna().sum()

person_age                       0
person_income                    0
person_home_ownership            0
person_emp_length              895
loan_intent                      0
loan_grade                       0
loan_amnt                        0
loan_int_rate                 3116
loan_status                      0
loan_percent_income              0
cb_person_default_on_file        0
cb_person_cred_hist_length       0
dtype: int64

In [281]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32581 entries, 0 to 32580
Data columns (total 12 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   person_age                  32581 non-null  int64  
 1   person_income               32581 non-null  int64  
 2   person_home_ownership       32581 non-null  object 
 3   person_emp_length           31686 non-null  float64
 4   loan_intent                 32581 non-null  object 
 5   loan_grade                  32581 non-null  object 
 6   loan_amnt                   32581 non-null  int64  
 7   loan_int_rate               29465 non-null  float64
 8   loan_status                 32581 non-null  int64  
 9   loan_percent_income         32581 non-null  float64
 10  cb_person_default_on_file   32581 non-null  object 
 11  cb_person_cred_hist_length  32581 non-null  int64  
dtypes: float64(3), int64(5), object(4)
memory usage: 3.0+ MB


In [282]:
mean_pel = df[df['person_emp_length'].isna() != True]['person_emp_length'].mean()
mean_lir = df[df['loan_int_rate'].isna() != True]['loan_int_rate'].mean()

print(mean_pel)
print(mean_lir)

4.789686296787225
11.01169489224439


In [283]:
min(df['loan_int_rate'])
max(df['loan_int_rate'])

23.22

In [284]:
df['loan_int_rate'] = df['loan_int_rate'].fillna(mean_lir)
df['person_emp_length'] = df['person_emp_length'].fillna(mean_pel)

In [285]:
df.isna().sum()

person_age                    0
person_income                 0
person_home_ownership         0
person_emp_length             0
loan_intent                   0
loan_grade                    0
loan_amnt                     0
loan_int_rate                 0
loan_status                   0
loan_percent_income           0
cb_person_default_on_file     0
cb_person_cred_hist_length    0
dtype: int64

In [286]:
df.head()

person_age  person_income person_home_ownership  person_emp_length  \
0          22          59000                  RENT              123.0   
1          21           9600                   OWN                5.0   
2          25           9600              MORTGAGE                1.0   
3          23          65500                  RENT                4.0   
4          24          54400                  RENT                8.0   

  loan_intent loan_grade  loan_amnt  loan_int_rate  loan_status  \
0    PERSONAL          D      35000          16.02            1   
1   EDUCATION          B       1000          11.14            0   
2     MEDICAL          C       5500          12.87            1   
3     MEDICAL          C      35000          15.23            1   
4     MEDICAL          C      35000          14.27            1   

   loan_percent_income cb_person_default_on_file  cb_person_cred_hist_length  
0                 0.59                         Y                           3  
1                 0.10                         N                           2  
2                 0.57                         N                           3  
3                 0.53                         N                           2  
4                 0.55                         Y                           4

In [287]:
df.loan_grade.value_counts(True)

A    0.330776
B    0.320770
C    0.198214
D    0.111292
E    0.029588
F    0.007397
G    0.001964
Name: loan_grade, dtype: float64

In [288]:
df.cb_person_default_on_file.value_counts(True)

N    0.82367
Y    0.17633
Name: cb_person_default_on_file, dtype: float64

In [289]:
min(df.person_age)
max(df.person_age)

144

In [290]:
df[df.person_age > 75]

person_age  person_income person_home_ownership  person_emp_length  \
81            144         250000                  RENT                4.0   
183           144         200000              MORTGAGE                4.0   
575           123          80004                  RENT                2.0   
747           123          78000                  RENT                7.0   
32297         144        6000000              MORTGAGE               12.0   
32355          78          48000                  RENT               41.0   
32416          94          24000                  RENT                1.0   
32422          80          64000                  RENT                7.0   
32506          84          94800              MORTGAGE                2.0   
32534          76          75000                  RENT               23.0   

      loan_intent loan_grade  loan_amnt  loan_int_rate  loan_status  \
81        VENTURE          C       4800      13.570000            0   
183     EDUCATION          B       6000      11.860000            0   
575     EDUCATION          B      20400      10.250000            0   
747       VENTURE          B      20000      11.011695            0   
32297    PERSONAL          C       5000      12.730000            0   
32355     MEDICAL          A       3000       7.510000            0   
32416     MEDICAL          C       6500      11.011695            0   
32422     MEDICAL          B       6800      11.710000            0   
32506    PERSONAL          A      10000       7.510000            0   
32534     MEDICAL          A      15000       8.940000            0   

       loan_percent_income cb_person_default_on_file  \
81                    0.02                         N   
183                   0.03                         N   
575                   0.25                         N   
747                   0.26                         N   
32297                 0.00                         N   
32355                 0.06                         N   
32416                 0.27                         N   
32422                 0.11                         N   
32506                 0.11                         N   
32534                 0.20                         N   

       cb_person_cred_hist_length  
81                              3  
183                             2  
575                             3  
747                             4  
32297                          25  
32355                          25  
32416                          27  
32422                          25  
32506                          24  
32534                          25

In [291]:
df = df[df.person_age <= 100]
len(df)

32576

In [292]:
len(df[df.loan_int_rate > 15])
# len(df[df.loan_int_rate < 10])

3441

In [304]:
df['int_group'] = np.where(df.loc[:, 'loan_int_rate'] < 10, 'under 10%', 
                           (np.where((df.loc[:, 'loan_int_rate'] >= 10) & (df.loc[:, 'loan_int_rate'] <= 15), '10% - 15%', 'over 15%')))


prettyTable(df, 'int_group')


Empty DataFrame
Columns: []
Index: [(10% - 15%, 54.17%, 17647), (under 10%, 35.27%, 11488), (over 15%, 10.56%, 3441)]

In [305]:
df['age_group'] = np.where(df['person_age'] <25, 'under 25', 
                           (np.where((df['person_age'] >= 25) & (df['person_age'] <35), '25 - 35', 'over 35')))


prettyTable(df, 'age_group')


Empty DataFrame
Columns: []
Index: [(25 - 35, 49.67%, 16180), (under 25, 37.8%, 12315), (over 35, 12.53%, 4081)]

In [306]:
df['risk_group'] = np.where((df['loan_grade'] == 'A') | (df['loan_grade'] == 'B'), 'low risk', 
                           (np.where((df['loan_grade'] == 'C') | (df['loan_grade'] == 'D'), 'medium risk', 'high risk')))


prettyTable(df, 'risk_group')


Empty DataFrame
Columns: []
Index: [(low risk, 65.16%, 21225), (medium risk, 30.95%, 10082), (high risk, 3.9%, 1269)]

In [307]:
df.person_income.max()

df[df['person_income'] == df.person_income.max()]
df[df['person_income'] == df.person_income.min()]


df['income_group'] = np.where((df['person_income'] < 65000), 'under 65k', 
                           (np.where((df['person_income'] >= 65000) & (df['person_income'] <= 125000), '65k - 125k', '> 125k')))


prettyTable(df, 'income_group')


Empty DataFrame
Columns: []
Index: [(under 65k, 61.13%, 19913), (65k - 125k, 32.33%, 10531), (> 125k, 6.54%, 2132)]

In [308]:
df.head()

person_age  person_income person_home_ownership  person_emp_length  \
0          22          59000                  RENT              123.0   
1          21           9600                   OWN                5.0   
2          25           9600              MORTGAGE                1.0   
3          23          65500                  RENT                4.0   
4          24          54400                  RENT                8.0   

  loan_intent loan_grade  loan_amnt  loan_int_rate  loan_status  \
0    PERSONAL          D      35000          16.02            1   
1   EDUCATION          B       1000          11.14            0   
2     MEDICAL          C       5500          12.87            1   
3     MEDICAL          C      35000          15.23            1   
4     MEDICAL          C      35000          14.27            1   

   loan_percent_income cb_person_default_on_file  cb_person_cred_hist_length  \
0                 0.59                         Y                           3   
1                 0.10                         N                           2   
2                 0.57                         N                           3   
3                 0.53                         N                           2   
4                 0.55                         Y                           4   

   int_group age_group   risk_group income_group  
0   over 15%  under 25  medium risk    under 65k  
1  10% - 15%  under 25     low risk    under 65k  
2  10% - 15%   25 - 35  medium risk    under 65k  
3   over 15%  under 25  medium risk   65k - 125k  
4  10% - 15%  under 25  medium risk    under 65k

In [309]:
prettyTable(df, 'loan_intent')

Empty DataFrame
Columns: []
Index: [(EDUCATION, 19.8%, 6451), (MEDICAL, 18.64%, 6071), (VENTURE, 17.55%, 5717), (PERSONAL, 16.94%, 5520), (DEBTCONSOLIDATION, 16.0%, 5212), (HOMEIMPROVEMENT, 11.07%, 3605)]

In [310]:
prettyTable(df, 'loan_grade')

Empty DataFrame
Columns: []
Index: [(A, 33.08%, 10777), (B, 32.07%, 10448), (C, 19.82%, 6456), (D, 11.13%, 3626), (E, 2.96%, 964), (F, 0.74%, 241), (G, 0.2%, 64)]

In [311]:
prettyTable(df, 'person_home_ownership')

Empty DataFrame
Columns: []
Index: [(RENT, 50.48%, 16443), (MORTGAGE, 41.26%, 13442), (OWN, 7.93%, 2584), (OTHER, 0.33%, 107)]

In [312]:
min(df['person_emp_length'])
max(df['person_emp_length'])

123.0

In [313]:
df = df[df['person_emp_length'] < 50]
df.head()

person_age  person_income person_home_ownership  person_emp_length  \
1          21           9600                   OWN                5.0   
2          25           9600              MORTGAGE                1.0   
3          23          65500                  RENT                4.0   
4          24          54400                  RENT                8.0   
5          21           9900                   OWN                2.0   

  loan_intent loan_grade  loan_amnt  loan_int_rate  loan_status  \
1   EDUCATION          B       1000          11.14            0   
2     MEDICAL          C       5500          12.87            1   
3     MEDICAL          C      35000          15.23            1   
4     MEDICAL          C      35000          14.27            1   
5     VENTURE          A       2500           7.14            1   

   loan_percent_income cb_person_default_on_file  cb_person_cred_hist_length  \
1                 0.10                         N                           2   
2                 0.57                         N                           3   
3                 0.53                         N                           2   
4                 0.55                         Y                           4   
5                 0.25                         N                           2   

   int_group age_group   risk_group income_group  
1  10% - 15%  under 25     low risk    under 65k  
2  10% - 15%   25 - 35  medium risk    under 65k  
3   over 15%  under 25  medium risk   65k - 125k  
4  10% - 15%  under 25  medium risk    under 65k  
5  under 10%  under 25     low risk    under 65k

In [314]:
min(df['loan_amnt'])
max(df['loan_amnt'])

35000

In [315]:
df['loan_size'] = np.where((df['loan_amnt'] < 10000), 'under 10k', 
                           (np.where((df['loan_amnt'] >= 10000) & (df['loan_amnt'] < 20000), '10k - 20k', '> 20k')))


prettyTable(df, 'loan_size')


/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


Empty DataFrame
Columns: []
Index: [(under 10k, 57.91%, 18862), (10k - 20k, 32.42%, 10560), (> 20k, 9.68%, 3152)]

In [316]:
df.loan_percent_income.min()
df.loan_percent_income.max()

0.83

In [317]:
df[df.loan_percent_income == df.loan_percent_income.max()]
df[df.loan_percent_income >= .8]
df[df.loan_percent_income == df.loan_percent_income.min()]
df[df.loan_percent_income == df.loan_percent_income.median()]

person_age  person_income person_home_ownership  person_emp_length  \
59             24          12000                   OWN                2.0   
103            25         230000              MORTGAGE                5.0   
250            22          14400                   OWN                0.0   
301            25         162000              MORTGAGE                5.0   
349            25         165000              MORTGAGE                0.0   
...           ...            ...                   ...                ...   
32541          52         163000              MORTGAGE                9.0   
32554          60          98000                  RENT                1.0   
32562          61         160000              MORTGAGE               30.0   
32577          54         120000              MORTGAGE                4.0   
32580          66          42000                  RENT                2.0   

             loan_intent loan_grade  loan_amnt  loan_int_rate  loan_status  \
59               VENTURE          E       1750      11.011695            0   
103              MEDICAL          C      35000      13.990000            0   
250             PERSONAL          B       2100      11.140000            0   
301    DEBTCONSOLIDATION          E      25000      16.000000            1   
349      HOMEIMPROVEMENT          F      25000      19.740000            0   
...                  ...        ...        ...            ...          ...   
32541           PERSONAL          D      25000      14.110000            1   
32554    HOMEIMPROVEMENT          D      15000      16.290000            1   
32562           PERSONAL          B      24000      10.380000            0   
32577           PERSONAL          A      17625       7.490000            0   
32580            MEDICAL          B       6475       9.990000            0   

       loan_percent_income cb_person_default_on_file  \
59                    0.15                         Y   
103                   0.15                         N   
250                   0.15                         N   
301                   0.15                         Y   
349                   0.15                         N   
...                    ...                       ...   
32541                 0.15                         N   
32554                 0.15                         N   
32562                 0.15                         N   
32577                 0.15                         N   
32580                 0.15                         N   

       cb_person_cred_hist_length  int_group age_group   risk_group  \
59                              3  10% - 15%  under 25    high risk   
103                             4  10% - 15%   25 - 35  medium risk   
250                             4  10% - 15%  under 25     low risk   
301                             3   over 15%   25 - 35    high risk   
349                             4   over 15%   25 - 35    high risk   
...                           ...        ...       ...          ...   
32541                          30  10% - 15%   over 35  medium risk   
32554                          29   over 15%   over 35  medium risk   
32562                          21  10% - 15%   over 35     low risk   
32577                          19  under 10%   over 35     low risk   
32580                          30  under 10%   over 35     low risk   

      income_group  loan_size  
59       under 65k  under 10k  
103         > 125k      > 20k  
250      under 65k  under 10k  
301         > 125k      > 20k  
349         > 125k      > 20k  
...            ...        ...  
32541       > 125k      > 20k  
32554   65k - 125k  10k - 20k  
32562       > 125k      > 20k  
32577   65k - 125k  10k - 20k  
32580    under 65k  under 10k  

[1256 rows x 17 columns]

In [320]:
df['loan_to_income_grouping'] = np.where((df['loan_percent_income'] < .1), 'under 10%', 
                           (np.where((df['loan_percent_income'] >= .1) & (df['loan_percent_income'] < .25), '10% - 25%', '> 25%')))


prettyTable(df, 'loan_to_income_grouping')


/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


Empty DataFrame
Columns: []
Index: [(10% - 25%, 51.04%, 16626), (under 10%, 27.47%, 8948), (> 25%, 21.49%, 7000)]

## EDA and Visuals

In [165]:
# groupings of diff categories by loan status
# hist of income, emp length

# loan status by age, by income level, by home ownnership
# by int rate amt, by loan amt

In [ ]:
# explore loan amnts 

In [194]:
df.columns.to_list()

['person_age',
 'person_income',
 'person_home_ownership',
 'person_emp_length',
 'loan_intent',
 'loan_grade',
 'loan_amnt',
 'loan_int_rate',
 'loan_status',
 'loan_percent_income',
 'cb_person_default_on_file',
 'cb_person_cred_hist_length',
 'age_group',
 'int_group',
 'risk_group',
 'income_group']

In [326]:
df.head(3)

person_age  person_income person_home_ownership  person_emp_length  \
1          21           9600                   OWN                5.0   
2          25           9600              MORTGAGE                1.0   
3          23          65500                  RENT                4.0   

  loan_intent loan_grade  loan_amnt  loan_int_rate  loan_status  \
1   EDUCATION          B       1000          11.14            0   
2     MEDICAL          C       5500          12.87            1   
3     MEDICAL          C      35000          15.23            1   

   loan_percent_income cb_person_default_on_file  cb_person_cred_hist_length  \
1                 0.10                         N                           2   
2                 0.57                         N                           3   
3                 0.53                         N                           2   

   int_group age_group   risk_group income_group  loan_size  \
1  10% - 15%  under 25     low risk    under 65k  under 10k   
2  10% - 15%   25 - 35  medium risk    under 65k  under 10k   
3   over 15%  under 25  medium risk   65k - 125k      > 20k   

  loan_to_income_grouping  
1               10% - 25%  
2                   > 25%  
3                   > 25%

In [323]:
prettyTable(df, 'loan_status')

Empty DataFrame
Columns: []
Index: [(0, 78.18%, 25467), (1, 21.82%, 7107)]

In [176]:
df['income_group'].groupby(df.loan_status).value_counts(True)

loan_status  income_group
0            under 65k       0.559901
             65k - 125k      0.364668
             > 125k          0.075431
1            under 65k       0.795413
             65k - 125k      0.175039
             > 125k          0.029548
Name: income_group, dtype: float64

In [177]:
df['loan_status'].groupby(df.income_group).value_counts(True)

income_group  loan_status
65k - 125k    0              0.881873
              1              0.118127
> 125k        0              0.901455
              1              0.098545
under 65k     0              0.716101
              1              0.283899
Name: loan_status, dtype: float64

In [182]:
df['age_group'].groupby(df.loan_status).value_counts(True)

loan_status  age_group
0            middle       0.502101
             young        0.371226
             old          0.126674
1            middle       0.477417
             young        0.402279
             old          0.120304
Name: age_group, dtype: float64

In [181]:
df['loan_status'].groupby(df.age_group).value_counts(True)

age_group  loan_status
middle     0              0.790297
           1              0.209703
old        0              0.790493
           1              0.209507
young      0              0.767806
           1              0.232194
Name: loan_status, dtype: float64

In [195]:
df['person_home_ownership'].groupby(df.loan_status).value_counts(True)

loan_status  person_home_ownership
0            MORTGAGE                 0.461421
             RENT                     0.441787
             OWN                      0.093886
             OTHER                    0.002906
1            RENT                     0.730407
             MORTGAGE                 0.237794
             OWN                      0.027156
             OTHER                    0.004643
Name: person_home_ownership, dtype: float64

In [196]:
df['loan_status'].groupby(df.person_home_ownership).value_counts(True)

person_home_ownership  loan_status
MORTGAGE               0              0.874265
                       1              0.125735
OTHER                  0              0.691589
                       1              0.308411
OWN                    0              0.925310
                       1              0.074690
RENT                   0              0.684284
                       1              0.315716
Name: loan_status, dtype: float64

In [197]:
df['loan_intent'].groupby(df.loan_status).value_counts(True)

loan_status  loan_intent      
0            EDUCATION            0.209683
             VENTURE              0.191189
             MEDICAL              0.174736
             PERSONAL             0.173636
             DEBTCONSOLIDATION    0.146150
             HOMEIMPROVEMENT      0.104606
1            MEDICAL              0.228085
             DEBTCONSOLIDATION    0.209652
             EDUCATION            0.156325
             PERSONAL             0.154355
             HOMEIMPROVEMENT      0.132405
             VENTURE              0.119178
Name: loan_intent, dtype: float64

In [198]:
df['loan_status'].groupby(df.loan_intent).value_counts(True)

loan_intent        loan_status
DEBTCONSOLIDATION  0              0.714121
                   1              0.285879
EDUCATION          0              0.827779
                   1              0.172221
HOMEIMPROVEMENT    0              0.738974
                   1              0.261026
MEDICAL            0              0.732993
                   1              0.267007
PERSONAL           0              0.801232
                   1              0.198768
VENTURE            0              0.851819
                   1              0.148181
Name: loan_status, dtype: float64

In [199]:
df['loan_grade'].groupby(df.loan_status).value_counts(True)

loan_status  loan_grade
0            A             0.381003
             B             0.343464
             C             0.200927
             D             0.058311
             E             0.013468
             F             0.002788
             G             0.000039
1            D             0.301112
             B             0.239341
             C             0.188406
             A             0.150978
             E             0.087379
             F             0.023920
             G             0.008864
Name: loan_grade, dtype: float64

In [201]:
df['loan_status'].groupby(df.loan_grade).value_counts(True)

loan_grade  loan_status
A           0              0.900427
            1              0.099573
B           0              0.837194
            1              0.162806
C           0              0.792596
            1              0.207404
D           1              0.590345
            0              0.409655
E           1              0.644191
            0              0.355809
F           1              0.705394
            0              0.294606
G           1              0.984375
            0              0.015625
Name: loan_status, dtype: float64

In [206]:
df['loan_size'].groupby(df.loan_status).value_counts(True)

loan_status  loan_size
0            under 10k    0.599364
             10k - 20k    0.318569
             > 20k        0.082067
1            under 10k    0.506261
             10k - 20k    0.344308
             > 20k        0.149430
Name: loan_size, dtype: float64

In [207]:
df['loan_status'].groupby(df.loan_size).value_counts(True)

loan_size  loan_status
10k - 20k  0              0.768277
           1              0.231723
> 20k      0              0.663071
           1              0.336929
under 10k  0              0.809246
           1              0.190754
Name: loan_status, dtype: float64

In [208]:
df['int_group'].groupby(df.loan_status).value_counts(True)

loan_status  int_group
0            10% - 15%    0.540935
             under 10%    0.402325
             over 15%     0.056740
1            10% - 15%    0.544674
             over 15%     0.280709
             under 10%    0.174617
Name: int_group, dtype: float64

In [209]:
df['loan_status'].groupby(df.int_group).value_counts(True)

int_group  loan_status
10% - 15%  0              0.780643
           1              0.219357
over 15%   1              0.579942
           0              0.420058
under 10%  0              0.891965
           1              0.108035
Name: loan_status, dtype: float64

In [210]:
df['risk_group'].groupby(df.loan_status).value_counts(True)

loan_status  risk_group 
0            low risk       0.724467
             medium risk    0.259237
             high risk      0.016296
1            medium risk    0.489517
             low risk       0.390319
             high risk      0.120163
Name: risk_group, dtype: float64

In [211]:
df['loan_status'].groupby(df.risk_group).value_counts(True)

risk_group   loan_status
high risk    1              0.672971
             0              0.327029
low risk     0              0.869299
             1              0.130701
medium risk  0              0.654895
             1              0.345105
Name: loan_status, dtype: float64

In [212]:
df['income_group'].groupby(df.loan_status).value_counts(True)

loan_status  income_group
0            under 65k       0.559901
             65k - 125k      0.364668
             > 125k          0.075431
1            under 65k       0.795413
             65k - 125k      0.175039
             > 125k          0.029548
Name: income_group, dtype: float64

In [213]:
df['loan_status'].groupby(df.income_group).value_counts(True)

income_group  loan_status
65k - 125k    0              0.881873
              1              0.118127
> 125k        0              0.901455
              1              0.098545
under 65k     0              0.716101
              1              0.283899
Name: loan_status, dtype: float64

In [321]:
df['loan_to_income_grouping'].groupby(df.loan_status).value_counts(True)

loan_status  loan_to_income_grouping
0            10% - 25%                  0.549967
             under 10%                  0.310794
             > 25%                      0.139239
1            > 25%                      0.486000
             10% - 25%                  0.368651
             under 10%                  0.145350
Name: loan_to_income_grouping, dtype: float64

In [322]:
df['loan_status'].groupby(df.loan_to_income_grouping).value_counts(True)

loan_to_income_grouping  loan_status
10% - 25%                0              0.842415
                         1              0.157585
> 25%                    0              0.506571
                         1              0.493429
under 10%                0              0.884555
                         1              0.115445
Name: loan_status, dtype: float64

In [324]:
df.columns.to_list()

['person_age',
 'person_income',
 'person_home_ownership',
 'person_emp_length',
 'loan_intent',
 'loan_grade',
 'loan_amnt',
 'loan_int_rate',
 'loan_status',
 'loan_percent_income',
 'cb_person_default_on_file',
 'cb_person_cred_hist_length',
 'int_group',
 'age_group',
 'risk_group',
 'income_group',
 'loan_size',
 'loan_to_income_grouping']

**The lower two income groups see higher rates of default, and make up a greater share of the defaults. In particular, the under 65k group sees a much higher level of risk**

    * 80% of all defaults are within the under 65k group
    * 28% of the under 65k group defaulted

**73% of all defaults were in the Rent category (50% of records were renters)**

**The Other category and Rent category have a 31% default rate**

**Medical and Debt Consolidation make up the highest portions of defaults at 23% and 21% respectively. Debt Consolidation also has the highest Default rate among the Loan Intent categories. This makes sense given that debt consolidtion often takes place among those in difficult financial situations.**

**30% of all defaults were grade D, and 24% were grade B. C and D together make up roughly half of all defaults. Grades D, E, F, and G all had default rates of 59% or higher. This makes sense since lower grades equate to higher risk. Higher default rates are expected within these groups. Conversely, grades A, B, and C see default rates of 10%, 16%, and 21% respectively.**

**The larger the loan amount, the higher the rate of default. A higher borrowing value means higher payments and therefore higher risk for the borrower.**